File: iban-validator/src/test/java/de/rwth/swc/banking/TestConfig.java
package de.rwth.swc.banking;

import de.rwth.swc.interact.integrator.rest.RestTemplateIntegrationInterceptor;
import de.rwth.swc.interact.integrator.rest.TestRestTemplateIntegrationInterceptor;
import de.rwth.swc.interact.observer.rest.RestTemplateObservationInterceptor;
import de.rwth.swc.interact.observer.rest.TestRestTemplateObservationInterceptor;
import org.springframework.boot.test.context.TestConfiguration;
import org.springframework.boot.test.web.client.TestRestTemplate;
import org.springframework.boot.web.client.RestTemplateBuilder;
import org.springframework.context.annotation.Bean;

@TestConfiguration
public class TestConfig {

    @Bean
    public TestRestTemplate testRestTemplate(
            RestTemplateBuilder restTemplateBuilder,
            TestRestTemplateObservationInterceptor interceptor1,
            TestRestTemplateIntegrationInterceptor interceptor2
    ) {
        var testRestTemplate = new TestRestTemplate(restTemplateBuilder);
        testRestTemplate.getRestTemplate().getInterceptors()
                .removeIf(it -> it instanceof RestTemplateObservationInterceptor || it instanceof RestTemplateIntegrationInterceptor);


        testRestTemplate.getRestTemplate().getInterceptors().add(interceptor2);
        testRestTemplate.getRestTemplate().getInterceptors().add(interceptor1);
        return testRestTemplate;
    }
}

File: iban-validator/src/test/kotlin/de/rwth/swc/banking/TestConfig.kt
package de.rwth.swc.banking

import de.rwth.swc.interact.rest.observer.WebClientObserver
import org.springframework.boot.test.context.TestConfiguration
import org.springframework.boot.web.servlet.context.ServletWebServerApplicationContext
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Primary
import org.springframework.web.reactive.function.client.WebClient
import org.springframework.web.reactive.function.client.support.WebClientAdapter
import org.springframework.web.service.invoker.HttpServiceProxyFactory


@TestConfiguration
class TestConfig {

    @Bean
    @Primary
    fun blacklistCheckTestApi(builder: WebClient.Builder): BlacklistCheckerApi {
        val webClient = builder
            .baseUrl("http://localhost:8082")
            .filter(WebClientObserver(false))
            .build()
        val httpServiceProxyFactory = HttpServiceProxyFactory
            .builder(WebClientAdapter.forClient(webClient))
            .build()
        return httpServiceProxyFactory.createClient(BlacklistCheckerApi::class.java)
    }

    companion object {
        fun ibanValidatorApi(
            builder: WebClient.Builder,
            port: Int
        ): IbanValidationApi {
            val webClient = builder
                .baseUrl("http://localhost:$port")
                .filter(WebClientObserver(true))
                .build()
            val httpServiceProxyFactory = HttpServiceProxyFactory
                .builder(WebClientAdapter.forClient(webClient))
                .build()
            return httpServiceProxyFactory.createClient(IbanValidationApi::class.java)
        }
    }
}
